drivers: hwinfo: add GD32F4xx support

meshium opened the pull request:
## Summary
Add HWINFO support for GD32F4xx:
- device unique ID (96-bit UID) readout
- reset cause reporting via `RCU_RSTSCK` (and clear support when available)

## Testing
- Twister (build/test):
  - `./scripts/twister -T tests/drivers/hwinfo/api -p <board> -v`
- Direct build:
  - `west build -b <board> -p always tests/drivers/hwinfo/api`

meshium committed:
drivers: hwinfo: add GD32 support

Add a hwinfo backend for GD32 providing:
- device unique ID readout from the 96-bit UID registers
- reset cause reporting via the RCU_RSTSCK reset status register

Reset cause mapping uses the hal_gigadevice RCU_RSTSCK flag definitions
and reports supported causes based on the macros available in the HAL.

Signed-off-by: Aleksandr Senin <al@meshium.net>

egilll commented:
- It seems you have a few incorrect flags copied over from STM32, the flags `IWDGRSTF`, `WWDGRSTF` and `PINRSTF` do not exist on GD32 (the correct equivalent of `WWDGRSTF` is `WWDG**T**RSTF`)
- In [this](https://github.com/egilll/zephyr-gd32/commit/82e0165840aea59b86217aa539eca6d5562a8306) commit I've fixed the above issue and added support for the other GD32 chips as well. Feel free to squash that into this pull request, otherwise I can make another pull request later.

meshium commented:
> * It seems you have a few incorrect flags copied over from STM32, the flags `IWDGRSTF`, `WWDGRSTF` and `PINRSTF` do not exist on GD32 (the correct equivalent of `WWDGRSTF` is `WWDG**T**RSTF`)
> * In [this](https://github.com/egilll/zephyr-gd32/commit/82e0165840aea59b86217aa539eca6d5562a8306) commit I've fixed the above issue and added support for the other GD32 chips as well. Feel free to squash that into this pull request, otherwise I can make another pull request later.

You’re right — thanks for catching that. I’ll cherry-pick and squash it into this PR (including the broader GD32 support).

meshium commented on drivers/hwinfo/Kconfig.gd32:
```diff
@@ -0,0 +1,18 @@
+# Copyright The Zephyr Project Contributors
```
Updated — I was confused by other Kconfig files.

soburi commented on soc/gd/gd32/gd32a50x/gd32_regs.h:
```diff
@@ -9,12 +9,12 @@
-#define RCU_APB2EN_OFFSET    0x18U
-#define RCU_CFG1_OFFSET      0x2CU
-#define RCU_CFG2_OFFSET      0x30U
+#define RCU_CFG0_OFFSET   0x04U
```
Do not include diff that are not related to the main body of the fix.

alexanderwachter commented on drivers/hwinfo/hwinfo_gd32.c:
```diff
@@ -0,0 +1,176 @@
+int z_impl_hwinfo_get_reset_cause(uint32_t *cause)
+{
+	uint32_t flags = 0U;
+	uint32_t rsts = RCU_RSTSCK;
```
I'm not familiar with the GD hal, but does that really read the cause from a register?
I would have expected some dereferencing here.

meshium replied:
Yes, it does read the register. In the GD32 HAL, `RCU_RSTSCK` is defined as `REG32(RCU + 0x24)`, and `REG32` already dereferences the address:

`#define REG32(addr)  (*(volatile uint32_t *)(uint32_t)(addr))`

So no extra dereferencing is needed.
